### PR TITLE
Draft: TLV-JSON-TLV converter

### PR DESCRIPTION
**Pull request by chrisdecenzo**
(for message serialization between android tv-app platform and content apps)

TODO:
- Add readme describing JSON schema
- Add call from android casting app
- Add translation to/from android intent to android tv-app

### CODE REVIEW

File: examples/tv-casting-app/tv-casting-common/src/CastingServer.cpp [hunk 1]
```diff
@@ -139,6 +139,7 @@ CHIP_ERROR CastingServer::OpenBasicCommissioningWindow(std::function<void(CHIP_E
     mOnConnectionSuccessClientCallback = onConnectionSuccess;
     mOnConnectionFailureClientCallback = onConnectionFailure;
     mOnNewOrUpdatedEndpoint            = onNewOrUpdatedEndpoint;
+    ChipLogError(AppServer, "-----------------------CastingServer PrepareForCommissioning");
```
Comment (chrisdecenzo) [inline examples/tv-casting-app/tv-casting-common/src/CastingServer.cpp:142]: TODO: remove

File: examples/common/json/JsonTLV.h [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+ */
+namespace Json {
+
+struct JsonContext
+{
+    Encoding::BufferWriter & mWriter;
+    void * mContext;
+};
+
+extern CHIP_ERROR WriteJSON(const TLVReader & aReader, Encoding::BufferWriter & bWriter);
```
Comment (andy31415) [inline examples/common/json/JsonTLV.h:59]: we probably do not need "extern"

File: examples/common/json/JsonTLV.h [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        errorMarker += "^";
+        ChipLogError(chipTool, "%s", errorMarker.c_str());
+        errorColumn.ClearValue();
+    }
+};
+
+class JsonTLVParser
+{
+public:
+    static CHIP_ERROR Put(chip::TLV::TLVWriter * writer, chip::TLV::Tag tag, Json::Value & value)
```
Comment (andy31415) [inline examples/common/json/JsonTLV.h:249]: This looks like complex code in a header. Should probably be in a C++ file.

Also please add unit tests.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Request changes: please add unit tests for such large code changes that are quite stand-alone (parsing and creating a JSON seems like it would be unit testable ... and unit tests would also explain how the JSON looks like)

File: examples/tv-casting-app/linux/CastingShellCommands.cpp [hunk 4]
```diff
@@ -80,6 +81,149 @@ void PrintBindings()
...
+
+} mockCommandSenderDelegate;
+
+CHIP_ERROR SendClusterCommand(chip::NodeId node, chip::EndpointId endpointId, chip::ClusterId clusterId, chip::CommandId commandId,
+                              uint16_t timedRequestTimeoutMs, uint16_t imTimeoutMs, char * json)
+{
+    chip::TLV::TLVWriter writer;
+    chip::TLV::TLVReader reader;
+
+    uint8_t buf[74];
```
Comment (andy31415) [inline examples/tv-casting-app/linux/CastingShellCommands.cpp:125]: Need to explain the sizes here ...74 seems a very particular constant.

File: examples/tv-casting-app/linux/CastingShellCommands.cpp [hunk 5]
```diff
@@ -80,6 +81,149 @@ void PrintBindings()
...
+
+    Json::Value outputValue;
+    VerifyOrReturnLogError(JsonParser2::ParseCustomArgument(testName, (char *) out_buf, outputValue), CHIP_ERROR_INVALID_ARGUMENT);
+
+    VerifyOrReturnLogError(inputValue == outputValue, CHIP_ERROR_DECODE_FAILED);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR RunJsonTests()
```
Comment (andy31415) [inline examples/tv-casting-app/linux/CastingShellCommands.cpp:195]: These tests should be unit tests run during test execution, not a shell command (we can include it in the casting app as well, but a variant of unit tests that are stand alone should exist).

**Comment by andy31415** [comment]
@chrisdecenzo I moved your PR to draft since the title said so.

File: examples/common/json/JsonTLV.cpp [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+
+    temp.Init(aReader);
+    tagControl = static_cast<TLVTagControl>(temp.GetControlByte() & kTLVTagControlMask);
+
+    for (size_t i = 0; i < aDepth; i++)
+        FormattedWrite(aWriter, "%s", aIndent);
+
+    if (TLVTypeIsContainer(type))
+    {
+        if (tagControl == TLVTagControl::ContextSpecific)
```
Comment (bzbarsky-apple) [inline examples/common/json/JsonTLV.cpp:103]: What about profile tags in TLV lists?

File: examples/common/json/JsonTLV.cpp [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+            FormattedWrite(aWriter, "[\n");
+            break;
+
+        default:
+            break;
+        }
+    }
+    else
+    {
+        if (tagControl == TLVTagControl::ContextSpecific)
```
Comment (bzbarsky-apple) [inline examples/common/json/JsonTLV.cpp:129]: Again, this does not seem to handle profile tags right.  But also, shouldn't that `"tag" : ` prefix just be written out independently of whether the tag is a container or not?

File: examples/common/json/JsonTLV.cpp [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+        case kTLVType_UTF8String:
+            err = temp.GetDataPtr(strbuf);
+            VerifyOrExit(err == CHIP_NO_ERROR, FormattedWrite(aWriter, "Error in kTLVType_UTF8String"));
+            FormattedWrite(aWriter, "\"%-.*s\"", static_cast<int>(len), strbuf);
+            break;
+
+        case kTLVType_ByteString:
+            err = temp.GetDataPtr(strbuf);
+            VerifyOrExit(err == CHIP_NO_ERROR, FormattedWrite(aWriter, "Error in kTLVType_ByteString"));
+            FormattedWrite(aWriter, "hex:");
```
Comment (bzbarsky-apple) [inline examples/common/json/JsonTLV.cpp:178]: I don't think this produces valid JSON, unless I am missing something.....

File: examples/common/json/JsonTLV.cpp [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+            err = temp.Get(sVal);
+            VerifyOrExit(err == CHIP_NO_ERROR, FormattedWrite(aWriter, "Error in kTLVType_SignedInteger"));
+            FormattedWrite(aWriter, "%" PRIi64, sVal);
+            break;
+
+        case kTLVType_UnsignedInteger:
+            uint64_t uVal;
+            err = temp.Get(uVal);
+            VerifyOrExit(err == CHIP_NO_ERROR, FormattedWrite(aWriter, "Error in kTLVType_UnsignedInteger"));
+            FormattedWrite(aWriter, "%" PRIu64, uVal);
```
Comment (bzbarsky-apple) [inline examples/common/json/JsonTLV.cpp:152]: We're OK with the data corruptions for values > 2^53?

File: examples/common/json/JsonTLV.h [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+ *
+ * Takes a functor to allocate the buffer to use for the hex bytes.  The functor
+ * is expected to return uint8_t *.  The caller is responsible for cleaning up
+ * this buffer as needed.
+ *
+ * On success, *octetCount is filled with the number of octets placed in the
+ * buffer.  On failure, the value of *octetCount is undefined.
+ */
+template <typename F>
+CHIP_ERROR ConvertHexToBytes(chip::CharSpan hex, F bufferAllocator, size_t * octetCount)
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply examples/common/json/JsonTLV.h:81]: We have HexToBytes too, inside the BytesToHex header.

File: examples/common/json/JsonTLV.h [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+
+    *octetCount = byteCount;
+    return CHIP_NO_ERROR;
+}
+
+class JsonParser2
+{
+public:
+    // Returns whether the parse succeeded.
+    static bool ParseComplexArgument(const char * label, const char * json, Json::Value & value)
```
Comment (bzbarsky-apple) [inline examples/common/json/JsonTLV.h:115]: This seems to be duplicating the chip-tool complex argument parser, right?  Is it different from that code?  If so, maybe it should not be called ParseComplexArgument...  If it's not different, why is it separate?

File: examples/common/json/JsonTLV.h [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+    {
+        chip::TLV::TLVType outer;
+
+        // need to check if its a tlvType or type
+        if (value.isMember("tlvType") && value.isMember("value"))
+        {
+            constexpr const char kHexNumPrefix[] = "0x";
+            constexpr size_t kHexNumPrefixLen    = ArraySize(kHexNumPrefix) - 1;
+
+            Json::Value objTlvType = value.get("tlvType", "0x0c");
```
Comment (bzbarsky-apple) [inline examples/common/json/JsonTLV.h:320]: Please document why the defaulting to 0x0c.

File: examples/common/json/JsonTLV.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/common/json/JsonTLV.h:81]: Well.... that is confusing ... maybe we can find a better name or split headers.

**Comment by yunhanw-google** [comment]
@chrisdecenzo  we are preparing one pull request that provide the generic TLV-JSON-TLV converter using C++ along with documentation, similar to the kotlin one we provided in https://github.com/project-chip/connectedhomeip/pull/26527
@emargolis is working on this c++ version, which should be ready within 1 or 2 weeks, you can consider to leverage that version

**Comment by andy31415** [comment]
This seems in draft for over 1 year. @chrisdecenzo can we close it?

**Comment by andy31415** [comment]
Closing very old PR in draft form.

**Pull request closed without merging**